### PR TITLE
JAMES-2809 BlobStoreVault contract for searching

### PR DESCRIPTION
**Pull request by trantienduchn**

### CODE REVIEW

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/metadata/DeletedMessageMetadataVault.java [hunk 1]
```diff
@@ -25,7 +25,7 @@
 import org.reactivestreams.Publisher;
 
 public interface DeletedMessageMetadataVault {
-    Publisher<Void> store(DeletedMessageWithStorageInformation deletedMessage);
+    Publisher<Void> store(User user, DeletedMessageWithStorageInformation deletedMessage);
```
Comment (chibenwa) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/metadata/DeletedMessageMetadataVault.java]: What happens when this user is different than the user field of the deletedMessage?

Looks like you are doing an API change to cope with a test logic issue.

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/blob/BlobStoreDeletedMessageVault.java [hunk 2]
```diff
@@ -79,6 +79,7 @@ public BlobStoreDeletedMessageVault(DeletedMessageMetadataVault messageMetadataV
         return Flux.from(messageMetadataVault.listRelatedBuckets())
             .concatMap(bucketName -> Flux.from(messageMetadataVault.listMessages(bucketName, user)))
             .map(DeletedMessageWithStorageInformation::getDeletedMessage)
+            .filter(deletedMessage -> user.equals(deletedMessage.getOwner()))
```
Comment (chibenwa) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/blob/BlobStoreDeletedMessageVault.java]: When does `Flux.from(messageMetadataVault.listMessages(bucketName, user)` returns things not related to the user?

I'd love to have a test explaining me that change...

File: mailbox/plugin/deleted-messages-vault/src/test/java/org/apache/james/vault/blob/BlobStoreDeletedMessageVaultTest.java [hunk 3]
```diff
@@ -50,6 +52,16 @@ public DeletedMessageVault getVault() {
...
 
 
+    @Disabled("the BlobStoreDeletedMessageVaultTest.append(user, deletedMessage, inputStream)" +
+        "doesn't use user argument, it use the owner retrieved from deletedMessage for storing messages" +
+        "this test passes not the same user and owner in DeletedMessage to append()" +
+        "as the consequence, the filter cannot rely on user argument" +
+        "It raised a concern in append() where we pass user information in two different arguments")
+    @Test
+    @Override
+    public void searchForAnUserShouldNotReturnMessagesFromAnotherUser() {
```
Comment (chibenwa) [inline mailbox/plugin/deleted-messages-vault/src/test/java/org/apache/james/vault/blob/BlobStoreDeletedMessageVaultTest.java]: A quick debug shows that

```
user: User{localPart=dimitri, domainPart=Optional[Domain : apache.org]}

 AND 

deletedMessageUser User{localPart=bob, domainPart=Optional[Domain : apache.org]}
```

Well that is a concern... This should be forbidden to store Dimitri message in bob vault, isn't it?

Shouldn't we rather reject that operation? Or better: modify the DeletedMessageVault API to remove that duplication and make the API less dangerous to play with:

```
public interface DeletedMessageVault {
    Publisher<Void> append(DeletedMessage deletedMessage, InputStream mimeMessage);
}
```

This clearly seems like the way to go...

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/blob/BlobStoreDeletedMessageVault.java (hunk 2 above)
Comment (trantienduchn) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/blob/BlobStoreDeletedMessageVault.java]: yeah seem like I didn't read the code carefully

**Comment by trantienduchn** [comment]
close due to the alternative solution

**Pull request closed without merging**